Introduce minimal retry functionality as a core framework feature

fmbenhassine opened the pull request:
This PR introduces a minimal core retry feature in Spring Framework. It is inspired by Spring Retry, but redesigned and trimmed to the bare minimum to cover most cases.

The main entry point is the `RetryTemplate` class, which is designed in a similar fashion to other templates provided by Spring. The package is minimal on purpose to keep its usage as simple as possible (see examples in `RetryTemplateTests`). It focuses only on stateless retries for now, but can be extended with other retry operations as well.

fmbenhassine committed:
Introduce a minimal retry functionality as a core framework feature

This commit introduces a minimal core retry feature. It is inspired
by Spring Retry, but redesigned and trimmed to the bare minimum to
cover most cases.

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/RetryPolicy.java:
```diff
@@ -0,0 +1,33 @@
+	 * Return the maximum number of retry attempts.
+	 * @return the maximum number of retry attempts
+	 */
+	int getMaxAttempts();
```
How do you envision a user possibly being able to implement a retry policy conditionally based on the actual exception thrown?

fmbenhassine replied:
Yes, I don't exclude extending the contract of `RetryPolicy` with anything that we think essential for this first iteration. The list of exceptions to retry/exclude could be part of the retry policy contract.

By "can be extended with other retry operations" I was referring to stateful retries. Stateful retries require some sort of context (similar to the one currently in Spring Retry). But instead of overwhelming the `RetryOperations` interface with both stateless and stateful operations, I envision stateful retries to be defined in a specialization of `RetryOperations` (something like `StatefulRetryOperations` with additional methods dealing with a context). I am trying to follow the ISP here, as most users won't probably need stateful retires, and therefore should not be forced to implement those methods. Do you agree?

fmbenhassine replied:
I pushed a change set that enriches the retry policy with a method that allows user to provide a predicate on the exception and specify the retry condition. I believe a predicate is more flexible than two lists of exceptions to include/exclude as it is currently in Spring Retry.

sbrannen reviewed (changes requested):
Hi @fmbenhassine,

First and foremost, thanks for putting this PR together! 👍 

I took a quick look, added a few comments, and asked a few questions (as food for thought).

And we will also discuss the proposal within the team.

Cheers,

Sam

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,136 @@
+ * @see BackOffPolicy
+ * @see RetryListener
+ */
+public class RetryTemplate implements RetryOperations {
```
Makes sense. What about a builder (similar to [the one in Spring Retry](https://github.com/spring-projects/spring-retry?tab=readme-ov-file#imperative-example))? Since we have at least 3 configuration options, a builder would be better than having several constructors IMO, do you agree?

fmbenhassine replied:
Done. I added two constructors with the retry policy and backoff policy. In hindsight, I see no need for a builder for now.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryOperations.java:
```diff
@@ -0,0 +1,45 @@
+	 * Retry the given callback (according to the retry policy configured at the implementation level)
+	 * until it succeeds or eventually throw an exception if the retry policy is exhausted.
+	 * @param retryCallback the callback to call initially and retry if needed
+	 * @param <R>           the type of the callback's result
```
I will fix that. But shouldn't the code style check task catch that? It seems to pass currently with this violation in the code.

fmbenhassine replied:
Fixed.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryOperations.java:
```diff
@@ -0,0 +1,45 @@
+	 * @return the callback's result
+	 * @throws Exception if the retry policy is exhausted
+	 */
+	<R> @Nullable R execute(Callable<R> retryCallback) throws Exception;
```
The idea was to minimize the number of interfaces the user has to deal with. I did not see any added value of having a `RetryCallback` with a single method over a plain `Callable` object. But now that you [mentioned](https://github.com/spring-projects/spring-framework/pull/34716#discussion_r2028396733) the necessity to have a name or description of the callback for diagnosis purposes, I believe it is better to have a dedicated interface for it. (BTW, the `RetryCallback` from Spring Retry has a [label](https://github.com/spring-projects/spring-retry/blob/main/src/main/java/org/springframework/retry/RetryCallback.java#L41-L49), which can be used for logging). I will introduce such a callback.

Regarding the dedicated exception type, I agree as well (for consistency). I will update the PR accordingly.

As per the exception type, I believe it should be checked. The reason is that as a client of the template, I am expecting it to retry the code and eventually throw an exception if the retry policy is exhausted (there is no (need for a) recovery callback, the recovery is actually the code I am expecting to write in the catch block).

fmbenhassine replied:
I introduced a `RetryCallback` interface to avoid being tied to `Callable`. The callback also provides a method to specify a logical name for diagnosis purposes.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,136 @@
+		int maxAttempts = this.retryPolicy.getMaxAttempts();
+		while (attempts++ <= maxAttempts) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("Retry attempt #" + attempts);
```
As mentioned [previously](https://github.com/spring-projects/spring-framework/pull/34716#discussion_r2028877670), I agree. I will add a callback with a label that is used to distinguish callbacks in log messages.

fmbenhassine replied:
Done

fmbenhassine commented:
Thank you for the review, @sbrannen !

I added a couple of inline comments and I will update the PR accordingly.

fmbenhassine commented:
I updated the PR with the following changes:

- Add/Use a specific `RetryException` in the API instead of the generic `Exception`
- Add `RetryCallback` interface with a logical name for diagnostics purposes
- Improve `RetryPolicy` with the ability to specify which exceptions to retry based on a `Predicate<Exception>`

I find this new version to be more coherent and consistent with other templates in terms of structure and API. Looking forward to the team's thoughts on this!

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/RetryCallback.java:
```diff
@@ -0,0 +1,42 @@
+	 * business operations.
+	 * @return the name of the callback
+	 */
+	String getName();
```
Not requirement, but I wonder if we can use a `default` convenient implementation here to the `getClass().getName()`?
Or generated lambda class name is not very useful?

sbrannen replied:
We have `ClassUtils.isLambdaClass()`, so we might be able to use that and base the name of the callback on the _enclosing class_ of the lambda.

But, TBH, I'm not sure what the enclosing class of a lambda is. I'd have to experiment with that.

fmbenhassine replied:
After our discussion, there is no need to introduce another interface.

> Not requirement, but I wonder if we can use a default convenient implementation here to the getClass().getName()?

I agree, @artembilan . I have used that as a default implementation for now. But I will let the team decide if we should return null and adapt the log messages accordingly.

fmbenhassine replied:
Spring Retry passes the callback itself to the listener methods (just to get the label), which is not very convenient in my opinion. But I agree, the `RetryExecution` object can be enriched with the callback name to give some context about the execution if needed.

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/RetryOperations.java:
```diff
@@ -0,0 +1,45 @@
+	 * @throws RetryException thrown if the retry policy is exhausted. All attempt exceptions
+	 * should be added as suppressed exceptions to the final exception.
+	 */
+	<R> @Nullable R execute(RetryCallback<R> retryCallback) throws RetryException;
```
I'm not very good with nullness, but feel like this contract has to `<R extends @Nullable Object>`.
So, the usage of this method would depend on the provided callback implementation.
I mean that is no up to us to say that `execute()` might return null.
The null could be only as a result of `RetryCallback` call.

fmbenhassine replied:
Good point, @artembilan ! And thank you for the pointer @sbrannen !

I updated the PR accordingly.

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,188 @@
+	 * @param retryListener the retry listener to use. Must not be <code>null</code>.
+	 */
+	public void setRetryListener(RetryListener retryListener) {
+		Assert.notNull(retryListener, "Retry listener must not be null");
```
@sbrannen , @jhoeller , do we still do these asserts if we promote nullness?

sbrannen replied:
Sometimes, but not always. 

We can discuss the pros and cons offline.

fmbenhassine replied:
I kept the check for now. I will defer the decision to the team about conventions here.

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,188 @@
+		String callbackName = retryCallback.getName();
+		// initial attempt
+		try {
+			if (logger.isDebugEnabled()) {
```
Not important.
Can we use a `LogAccessor` API in Spring Core itself to make this code more clean?

fmbenhassine replied:
I think it is not (only) about formatting, but rather about replacing all code blocks (`if (logger.isDebugEnabled()) { logger.debug(...); }`) with a single statement. I believe this is better for readability. Even if `LogAccessor` is not currently used in Spring Framework itself, may be this is an opportunity to start using it? I personally like that API and think to start using it in Batch.

For now, I updated the PR to use `LogAccessor`, but if the team thinks it's not appropriate, then I will revert that change.

sbrannen replied:
> I think it is not (only) about formatting, but rather about replacing all code blocks (`if (logger.isDebugEnabled()) { logger.debug(...); }`) with a single statement.

You don't have to use `LogAccessor` to achieve that.

We achieve that in multiple places using only `LogMessage`.

For example:

https://github.com/spring-projects/spring-framework/blob/b699b65b40a0299d4845c2d0160ce1a0145d85d4/spring-beans/src/main/java/org/springframework/beans/factory/support/AbstractBeanFactory.java#L1746-L1748

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/support/MaxRetryAttemptsPolicy.java:
```diff
@@ -0,0 +1,69 @@
+
+		@Override
+		public boolean shouldRetry(Throwable throwable) {
+			return this.retryAttempts++ < maxRetryAttempts;
```
I'm not familiar with Spring Framework code style, but I find this way as useful:
```
MaxRetryAttemptsPolicy.this.maxRetryAttempts
```

fmbenhassine replied:
I updated the code with that style, and I will let the team decide what is appropriate for this.

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/support/PredicateRetryPolicy.java:
```diff
@@ -0,0 +1,68 @@
+	/**
+	 * A {@link RetryExecution} based on a predicate.
+	 */
+	public static class PredicateRetryPolicyExecution implements RetryExecution {
```
Why `static`?
Why do you allow to use this instance outside of the policy itself?
Why we cannot use `PredicateRetryPolicy.this.predicate` property and lock-in the usage of this class?

artembilan reviewed (changes requested):
This is great even if I have left so many questions.
Thanks

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/RetryExecution.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Strategy interface to define a retry execution.
+ *
+ * <p>Implementations do not need to be thread-safe.
```
... buy may be stateful?

fmbenhassine replied:
Yes, I updated the javadoc accordingly.

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/support/MaxRetryAttemptsPolicy.java:
```diff
@@ -0,0 +1,69 @@
+	 * Set the maximum number of retry attempts.
+	 * @param maxRetryAttempts the maximum number of retry attempts. Must be greater than zero.
+	 */
+	public void setMaxRetryAttempts(int maxRetryAttempts) {
```
We might want to support both options.

fmbenhassine replied:
I updated the code to support both options.

fmbenhassine commented on spring-core/src/test/java/org/springframework/core/retry/support/MaxRetryDurationPolicyTests.java:
```diff
@@ -0,0 +1,54 @@
+		Thread.sleep(400);
+		assertThat(retryExecution.shouldRetry(Mockito.any(Exception.class))).isTrue();
+		Thread.sleep(400);
+		assertThat(retryExecution.shouldRetry(Mockito.any(Exception.class))).isFalse();
```
Indeed there is no guarantee for this test to be stable, so I removed it for now. If there is a better way to test this, please let me know and I will update accordingly.

fmbenhassine commented:
Thank you for the review, @artembilan and @sbrannen !

I updated the PR accordingly and added a few comments. Please let me know if we need other updates.

The build of the PR is failing for tests that are not related to this change set.

artembilan commented on spring-core/src/main/java/org/springframework/core/retry/support/CompositeRetryListener.java:
```diff
@@ -0,0 +1,84 @@
+	 * @param listeners the delegate listeners to register
+	 */
+	public CompositeRetryListener(List<RetryListener> listeners) {
+		this.listeners = listeners;
```
This is not correct.
We have to create a copy here internally.
Otherwise, any mutate to that external collection would effect our component as well.
And there is chance of opposite, more dangerous logic: we modify `this.listeners` internally here and everything else using that external collection would be effected.
I remember having this error a lot in our projects when we used SonarQube.

artembilan commented on spring-core/src/test/java/org/springframework/core/retry/support/MaxRetryAttemptsPolicyTests.java:
```diff
@@ -0,0 +1,54 @@
+		RetryExecution retryExecution = retryPolicy.start();
+
+		// then
+		assertThat(retryExecution.shouldRetry(Mockito.any(Exception.class))).isTrue();
```
You still use `Mockito.any()` in these tests while it is wrong context for this API.
Here the simple `Mockito.mock()` should be enough.
Or might be even better from performance perspective to use some real exception instance instead of reflection 😄

fmbenhassine replied:
You are right, apologies I just missed that from our last review. Should be fixed now.

spencergibb commented:
👋🏻  Joining the conversation here as Spring Cloud is an advanced user of Spring Retry, so we would be interested in how the functionality compares to the Spring Retry project.

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/support/PredicateRetryPolicy.java:
```diff
@@ -0,0 +1,50 @@
+
+	/**
+	 * Start a new retry execution.
+	 * @return a fresh ready to be used
```
A fresh what?

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,178 @@
+	 * @throws RetryException thrown if the retry policy is exhausted. All attempt exceptions
+	 * are added as suppressed exceptions to the final exception.
+	 */
+	@Override
```
Is this Javadoc different than the Javadoc declared in the interface?

If not, you can remove the duplicated Javadoc.

fmbenhassine replied:
Yes, it's different. It doesn't mention "according to the retry policy configured at the implementation level" and also has more details about suppressed exceptions.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/support/MaxRetryAttemptsPolicy.java:
```diff
@@ -0,0 +1,88 @@
+
+		MaxRetryAttemptsPolicyExecution() {
+			this.retryAttempts = 0;
+		}
```
That's right. Removed.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryListener.java:
```diff
@@ -0,0 +1,62 @@
+	 * @param result the result of the callback
+	 * @param <T> the type of the result
+	 */
+	default <T> void onRetrySuccess(RetryExecution retryExecution, T result) {
```
I think this was copied from the previous interface in Spring Retry. Indeed, I see no benefit for this generic type here, so I removed it.

fmbenhassine commented on spring-core/src/main/java/org/springframework/core/retry/RetryTemplate.java:
```diff
@@ -0,0 +1,178 @@
+		String callbackName = retryCallback.getName();
+		// initial attempt
+		try {
+			logger.debug("About to execute callback '" + callbackName + "'");
```
Good catch. I updated the code to use the `Supplier` variants.

sbrannen commented on spring-core/src/main/java/org/springframework/core/retry/support/CompositeRetryListener.java:
```diff
@@ -0,0 +1,85 @@
+	 */
+	public void addListener(RetryListener listener) {
+		Assert.notNull(listener, "Retry listener must not be null");
+		this.listeners.add(listener);
```
Good catch! 👍 

I've fixed that locally.
